tests/lib/rules/no-constant-expressions.js

```diff
@@ -0,0 +1,63 @@
+/**
+ * @fileoverview Tests for the no-mixed-requires rule.
+ * @author Raphael Pigulla
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+var eslint = require("../../../lib/eslint"),
+    ESLintTester = require("eslint-tester");
+
+//------------------------------------------------------------------------------
+// Tests
+//------------------------------------------------------------------------------
+
+var eslintTester = new ESLintTester(eslint);
+eslintTester.addRuleTest("lib/rules/no-constant-expressions", {
+    valid: [
+        { code: "5", args: 1 },
+        { code: "a", args: 1 },
+        { code: "b && c", args: 1 },
+        { code: "b & 7", args: 1 },
+        { code: "b | c", args: 1 },
+        { code: "b === 3 && c !== 4", args: 1 },
+        { code: "a && b && c && d && e && f && g && h && i && j && k && l && m", args: 1 },
+        { code: "a === foo(a)", args: 1 },
+        { code: "(null)(5) + 9", args: 1},
+        { code: "(null)(5) && a === 3", args: 1},
+        { code: "(foo || bar)(5) + 9", args: 1},
+        { code: "(foo || bar)(5) && a === 3", args: 1},
+        { code: "foo() || bar()", args: 1 },
+        { code: "(a = 5) === 5", args: 1 },
+        { code: "3 === (a ? 2 : 3)", args: 1 },
+        { code: "a - 9 > -25", args: 1 },
+        { code: "2 < a && a < 2.1", args: 1 },
+        { code: "a >= 0", args: 1 },
+        { code: "200000 < a && a < 200000.0000001", args: 1 },
+        { code: "a > 19 || a + 1 > 0", args: 1 },
+        { code: "5 === (a || true)", args: 1 },
+        { code: "a === 'true'", args: 1 },
+        { code: "'long string' + 'another string'", args: 1 },
+        { code: "3 + (-a)", args: 1 }
+    ],
+    invalid: [
+        { code: "5 + 3", args: 1, errors: [{ message: "Expression always 8", type: "BinaryExpression" }] },
+        { code: "true && false", args: 1, errors: [{ message: "Expression always false", type: "LogicalExpression" }] },
+        { code: "a * 0 + 5;", args: 1, errors: [{ message: "Expression always 5", type: "BinaryExpression"}] },
+        { code: "a === a;", args: 1, errors: [{ message: "Expression always true", type: "BinaryExpression"}] },
```

Comment:
Yeah, I know that. That's why I use a polyfill or `typeof x === "number" && isNaN(x)`. In my opinion `x === x` is a non-straightforward way to detect for `NaN`, it's confusing if you've never seen it before, and it could theoretically break if the language was extended by new capabilities, such as overriding equality operator or adding new "nullish" values.
We could disable the detection of `x === x`, but the possible benefit it provides by catching typo errors together with forcing an extra clarity for code seems beneficial to me.
But I don't mind if it's changed, I don't feel strongly about it.
It's likely the rule won't be included anyway. :)